src/dmd/audit_log.d

```diff
@@ -0,0 +1,81 @@
+/**
+ * Generate text files for assistance in debugging dmd.
+ *
+ */
+
+module dmd.audit_log;
+import core.stdc.stdlib;
+import core.stdc.stdio;
+import dmd.root.rootobject;
+import dmd.globals;
+import dmd.ast_node;
+import dmd.hdrgen;
+import dmd.expression;
+import dmd.dsymbol;
+import dmd.statement;
+import dmd.mtype;
+import dmd.init;
+import dmd.dtemplate;
+import dmd.common.outbuffer;
+
+private FILE*[string] file;
+
+void Audit_Log(ASTNode node, string functioncall = __FUNCTION__, string modulecall = __MODULE__)
+{
+    debug(AUDITLOG)
+    {
+        bool Exist()
+        {
+            for(int x = 0; x < global.params.calls.length; x++)
+            {
+                if(global.params.calls[x].toString() == functioncall
+                   || global.params.calls[x].toString() ==  modulecall)
+                    return true;
+            }
+            return false;
+        }
+        if(!Exist())
+            return;
+        auto x = "debug/" ~ modulecall ~ "/" ~ functioncall ~ ".txt\0";
+        auto y = fopen(x.ptr, "w+\0".ptr);
+        file[modulecall ~ " " ~ functioncall] = y;
+        OutBuffer buf;
+        HdrGenState hgs;
+        switch(node.dyncast())
+        {
+            case DYNCAST.expression:
+                toCBuffer(cast(Expression) node, &buf, &hgs);
+                break;
+            case DYNCAST.statement:
+                toCBuffer(cast(Statement) node, &buf, &hgs);
+                break;
+            case DYNCAST.dsymbol:
+                toCBuffer(cast(Dsymbol) node, &buf, &hgs);
+                break;
+            case DYNCAST.initializer:
+                toCBuffer(cast(Initializer) node, &buf, &hgs);
+                break;
+            case DYNCAST.type:
+                auto t = cast(Type) node;
+                toCBuffer(t, &buf, t.getTypeInfoIdent(), &hgs);
+                break;
+            case DYNCAST.templateparameter:
+                toCBuffer(cast(TemplateParameter) node, &buf, &hgs);
+                break;
+            default:
+                break;
+        }
+        fprintf(file[modulecall ~ " " ~ functioncall], buf.extractChars());
```

Comment:
To avoid recomputing the hash key and performing a redundant hash lookup, it would be better to store the file handle in a local variable after the first time we look it up in the cache.

Reply:
Also, better to use `fputs`or `fprintf(file[...], "%s", ...)` here, to avoid a potential format-string injection vulnerability.